src/Sign.SignatureProviders.TrustedSigning/TrustedSigningService.cs

```diff
@@ -0,0 +1,99 @@
+        private readonly SemaphoreSlim _mutex = new(1);
+        private X509Certificate2? _publicKey;
+
+        public TrustedSigningService(
+            IServiceProvider serviceProvider,
+            TokenCredential tokenCredential,
+            Uri endpointUrl,
+            string accountName,
+            string certificateProfileName)
+        {
+            ArgumentNullException.ThrowIfNull(serviceProvider, nameof(serviceProvider));
+            ArgumentNullException.ThrowIfNull(tokenCredential, nameof(tokenCredential));
+            ArgumentNullException.ThrowIfNull(endpointUrl, nameof(endpointUrl));
+            ArgumentException.ThrowIfNullOrEmpty(accountName, nameof(accountName));
+            ArgumentException.ThrowIfNullOrEmpty(certificateProfileName, nameof(certificateProfileName));
+
+            _client = new CertificateProfileClient(tokenCredential, endpointUrl);
+            _accountName = accountName;
+            _certificateProfileName = certificateProfileName;
+
+            _logger = serviceProvider.GetRequiredService<ILogger<TrustedSigningService>>();
+        }
+
+        public void Dispose()
+        {
+            _mutex.Dispose();
+            _publicKey?.Dispose();
+            GC.SuppressFinalize(this);
+        }
+
+        public async Task<X509Certificate2> GetCertificateAsync(CancellationToken cancellationToken)
+        {
+            if (_publicKey is not null)
+            {
+                return new X509Certificate2(_publicKey);
+            }
+
+            await _mutex.WaitAsync();
+            try
+            {
+                if (_publicKey is null)
+                {
+                    Stopwatch stopwatch = Stopwatch.StartNew();
+
+                    _logger.LogTrace(Resources.FetchingCertificate);
+
+                    CertificateProfileSignOperation operation = await _client.StartSignAsync(_accountName, _certificateProfileName, _emptyRequest, cancellationToken: cancellationToken);
+                    Response<SignStatus> response = await operation.WaitForCompletionAsync(cancellationToken);
+
+                    byte[] rawData = Convert.FromBase64String(Encoding.UTF8.GetString(response.Value.SigningCertificate));
+                    X509Certificate2Collection collection = [];
+                    collection.Import(rawData);
+
+                    _publicKey = collection[collection.Count - 1];
+
+                    _logger.LogTrace(Resources.FetchedCertificate, stopwatch.Elapsed.TotalMilliseconds);
+                }
+            }
+            finally
+            {
+                _mutex.Release();
+            }
+
+            return new X509Certificate2(_publicKey);
+        }
+
+        public async Task<RSA> GetRsaAsync(CancellationToken cancellationToken)
+        {
+            X509Certificate2 publicKey = await GetCertificateAsync(cancellationToken);
+            return new RSATrustedSigning(_client, _accountName, _certificateProfileName, publicKey);
```

Comment:
Can this be reused as well?

Reply:
The object will be `Disposed` so we cannot reuse it.